auth/providers/azure/graph/graph.go

```diff
@@ -151,23 +182,162 @@ func (u *UserInfo) getExpandedGroups(ctx context.Context, ids []string) (*GroupL
+		}
+		return nil, errors.Wrap(err, "Error while parsing accessToken for validation")
+	}
+
+	// the arc obo service does not support getting groups for applications
+	if claims[idtypClaim] != nil {
+		return nil, errors.New("Overage claim (users with more than 200 group membership) for SPN is currently not supported. For troubleshooting, please refer to aka.ms/overageclaimtroubleshoot")
+	}
+
+	buf := new(bytes.Buffer)
+	if err := json.NewEncoder(buf).Encode(reqBody); err != nil {
+		return nil, errors.Wrap(err, "failed to encode token request")
+	}
+	endpoint, err := getOBORegionalEndpoint(u.region, u.resourceID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create getMemberGroups request")
+	}
+	req, err := http.NewRequest(http.MethodPost, endpoint, buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create getMemberGroups request")
+	}
+	// Set the auth headers
+	req.Header = u.headers
+
+	correlationID := uuid.New()
+	req.Header.Set("x-ms-correlation-request-id", correlationID.String())
+	getMemberGroupsCtx, cancel := context.WithTimeout(context.Background(), getMemberGroupsTimeout)
+	defer cancel()
+	start := time.Now()
+	klog.V(5).Infof("Sending getMemberGroups request with correlationID: %s", correlationID.String())
+	// use default httpclient without retries first
+	client := *httpclient.DefaultHTTPClient
+	resp, err := client.Do(req.WithContext(getMemberGroupsCtx))
+	duration := time.Since(start).Seconds()
+	internalServerCode := strconv.Itoa(http.StatusInternalServerError)
+	if err != nil {
+		getMemberGroupsUsingARCOboServiceCounter.WithLabelValues(internalServerCode).Inc()
+		getMemberGroupsUsingARCOboServiceHistogram.WithLabelValues(internalServerCode).Observe(duration)
+		klog.V(5).Infof("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups: %s", correlationID.String(), err.Error())
+		return nil, errors.Errorf("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups", correlationID.String())
+	}
+
+	// use retryable client only for unavailable and gatewaytimeout errors
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusGatewayTimeout {
+		start = time.Now()
+		resp, err = u.client.Do(req.WithContext(ctx))
+		duration = time.Since(start).Seconds()
+		if err != nil {
+			getMemberGroupsUsingARCOboServiceCounter.WithLabelValues(internalServerCode).Inc()
+			getMemberGroupsUsingARCOboServiceHistogram.WithLabelValues(internalServerCode).Observe(duration)
+			klog.V(5).Infof("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups on retries: %s", correlationID.String(), err.Error())
+			return nil, errors.Errorf("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups on retries", correlationID.String())
+		}
+	}
+
+	defer resp.Body.Close()
+	respStatusCode := strconv.Itoa(resp.StatusCode)
+	getMemberGroupsUsingARCOboServiceCounter.WithLabelValues(respStatusCode).Inc()
```

Comment:
It seems this should be moved before if condition to check response status.

Reply:
already discussed in previous comment